### PR TITLE
Convert diff view to ts

### PR DESCRIPTION
**Pull request by jeremy-rifkin**

### CODE REVIEW

File: static/components.js [hunk 1]
```diff
@@ -208,6 +208,20 @@ module.exports = {
...
+            componentState: {},
+        };
+    },
+    getDiffViewWith: function (lhs, rhs) {
+        return {
+            type: 'component',
+            componentName: 'diff',
+            componentState: {lhs: lhs, rhs: rhs},
+        };
+    },
```
Comment (matsjla) [inline static/components.js:224]: Yay I planned to do this but totally forgot, awesome!

File: static/panes/diff.ts [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+                            `<li class="compilerId">${escape(getItemDisplayTitle(item))}</li>` +
+                            '</ul>' +
+                            '</div>'
+                        );
+                    },
+                },
+                dropdownParent: 'body',
+                plugins: ['input_autogrow'],
+                onChange: value => {
+                    if (!((value as any as string) in this.compilers)) return;
```
Comment (matsjla) [inline static/panes/diff.ts:236]: Assume this is another case of value being `string | string[] | number`? I guess this is the best we can do

File: static/panes/diff.ts [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        // this gets called during the super's constructor before lhs/rhs have been initialized
+        if ((this.lhs as any) !== undefined && (this.rhs as any) !== undefined) {
+            if (this.lhs.compiler && this.rhs.compiler) {
+                return `${this.lhs.compiler.name} vs ${this.rhs.compiler.name}`;
+            }
+        }
+        return '';
+    }
+
+    updateCompilersFor(selectize: TomSelect, id: number | string) {
```
Comment (matsjla) [inline static/panes/diff.ts:397]: this.compilers is typed as `Record<string, ...>` doesn't that mean the number part of the union is wrong?

**Review by matsjla — COMMENTED** [review]
Code looks great, few tiny nits. I am yet to test, but happy overall! Awesome

File: static/panes/diff.ts [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+        options: unknown,
+        editorId: number,
+        treeId: number
+    ) {
+        if (!compiler) return;
+        options = options || '';
+        let name = compiler.name + ' ' + options;
+        // TODO: selectize doesn't play nicely with CSS tricks for truncation; this is the best I can do
+        // There's a plugin at: http://www.benbybenjacobs.com/blog/2014/04/09/no-wrap-plugin-for-selectize-dot-js
+        // but it doesn't look easy to integrate.
```
Comment (RubenRBS) [inline static/panes/diff.ts:346]: Is this still the case for TomSelect?

**Review by RubenRBS — APPROVED** [review]

File: static/panes/diff.ts (hunk 2 above)
Comment (jeremy-rifkin) [reply static/panes/diff.ts:236]: Yeah it's another case of the tom select typing weirdness, I'd like to find a better general solution for this...

File: static/panes/diff.ts (hunk 3 above)
Comment (jeremy-rifkin) [reply static/panes/diff.ts:397]: The diff view differentiates between executor and compiler ids by appending `_exec` to editors, I'll update the type to be `Record<string | number, ...>`

**Comment by jeremy-rifkin** [comment]
Thank you @RubenRBS for spotting the type improvement opportunities! Embarrassingly I didn't realize the `types/` directly existed 😄

File: static/panes/diff.ts (hunk 4 above)
Comment (jeremy-rifkin) [reply static/panes/diff.ts:329]: It appears so

**Comment by partouf** [comment]
Lgtm

**Review by matsjla — APPROVED** [review]
LGTM

**Pull request merged**